### ISSUE TITLE
Logic bug in parser function.

### ISSUE THREAD

**Issue by destr4ct** [acassen/keepalived#2440]
**Logic bug in parser function**
The `have_brace` function in lib/parser.c doesn't handle the case where a backslash is the last character in conf_file. In this situation, the loop skips the null-terminator and continues searching for a curly brace beyond the allocated chunk for the string, resulting in an out-of-bounds access.

```
static bool __attribute__((pure))have_brace(const char *conf_file){
  const char *p = conf_file;
  if (!*p)
    return false;
  do {
     if (*p == '\\')
       p++;                            // ptr on null-byte
     else if (*p == '{')
       return true;
  } while (*++p);                // skipping null-byte immediately!
  return false;
} 
```

**To Reproduce**
1. Configure keepalived with Address Sanitizer (ASAN)
2. Run keepalived with configuration file specified (`keepalived -l -D -n -C -f repro01-04-07-2024.txt`)

**Expected behavior**
The function should handle the case of a trailing backslash without accessing memory beyond the string's bounds.

**Keepalived version**
```
Keepalived v2.3.1 (07/01,2024), git commit v2.3.1-20-gcf0f3e53
```

**Distro**
Name: ubuntu
Version: 22.04
Architecture: aarch64

**Configuration file**
Provided as attachment

**Did keepalived coredump?**
No coredump, but ASAN reports a heap-buffer-overflow error.

**Error reported by ASAN**
```
==797025==ERROR: AddressSanitizer: heap-buffer-overflow on address 0xfffff2903680 at pc 0xaaaaaae3ecc8 bp 0xffffffffe300 sp 0xffffffffe310
READ of size 1 at 0xfffff2903680 thread T0
    #0 0xaaaaaae3ecc4 in have_brace /home/parallels/keepalived/lib/parser.c:1655
    #1 0xaaaaaae3ecc4 in open_and_check_glob /home/parallels/keepalived/lib/parser.c:1706
    #2 0xaaaaaae4369c in open_glob_file /home/parallels/keepalived/lib/parser.c:2369
    #3 0xaaaaaae48ae4 in check_include /home/parallels/keepalived/lib/parser.c:2508
    #4 0xaaaaaae48ae4 in read_line /home/parallels/keepalived/lib/parser.c:2840
    #5 0xaaaaaae5a144 in process_stream /home/parallels/keepalived/lib/parser.c:3002
    #6 0xaaaaaae5ee04 in init_data /home/parallels/keepalived/lib/parser.c:3251
    #7 0xaaaaaabd4ff0 in read_config_file /home/parallels/keepalived/keepalived/core/main.c:498
    #8 0xaaaaaabd4ff0 in keepalived_main /home/parallels/keepalived/keepalived/core/main.c:2660
    #9 0xfffff78b73f8 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #10 0xfffff78b74c8 in __libc_start_main_impl ../csu/libc-start.c:392
    #11 0xaaaaaab005ac in _start (/home/parallels/keepalived/bin/keepalived+0x605ac)
```

**Comment by destr4ct** [comment]
[repro01-04-07-2024.txt](https://github.com/user-attachments/files/16126774/repro01-04-07-2024.txt)

**Comment by pqarmitage** [comment]
@destr4ct The config file repro01-04-07-2024.txt attached above is not plain text. Could you please post a plain text version o your config.

**Comment by destr4ct** [comment]
@pqarmitage, the bug in code I've shown Is easy to spot and the configuration file is just a mess (syntax-incorrect, contains bad bytes) generated by AFL++ fuzzer with custom dict. I don't think there is any reason to minimize and clean it. I've pinned it just to help you with reproducing. 

I can code a patch on request.

**Comment by pqarmitage** [comment]
@destr4ct Many thanks for reporting this issue. This should now be resolved by commit 0b894f9.

Your testing is greatly appreciated, and any further bug reports are welcome.

**Comment by destr4ct** [comment]
Seems to be suitable, thanks!
